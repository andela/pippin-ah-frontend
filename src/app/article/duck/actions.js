import types from './types';

const setCreateStatus = createStatus => ({
  type: types.SET_CREATE_STATUS,
  createStatus,
});

const setFetchArticleState = fetchArticleState => ({
  type: types.SET_FETCH_ARTICLE_STATE,
  fetchArticleState,
});

const setFetchArticleError = errorMessage => ({
  type: types.SET_FETCH_ARTICLE_ERROR,
  errorMessage,
});

// const setArticleCategory = articleCategory => ({
//   type: types.SET_ARTICLE_CATEGORY,
//   articleCategory,
// });

const addArticleData = (articleData, articleCategory) => ({
  type: types.ADD_ARTICLE_DATA,
  articleData,
  articleCategory,
});

export default {
  setFetchArticleState,
  setFetchArticleError,
  // setArticleCategory,
  addArticleData,
  setCreateStatus,
};
