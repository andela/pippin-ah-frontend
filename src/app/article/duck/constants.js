export default {
  CREATING: 'CREATING',
  CREATE_ERROR: 'CREATE_ERROR',
  CREATE_SUCCESS: 'CREATE_SUCCESS',
  FETCH_SINGLE_SUCCESS: 'FETCH_SINGLE_SUCCESS',
  FETCH_SINGLE_ERROR: 'FETCH_SINGLE_ERROR',
  FETCHING_SINGLE: 'FETCHING_SINGLE',
  FETCHING_ARTICLE: 'FETCHING_ARTICLE',
  FETCH_ARTICLE_ERROR: 'FETCH_ARTICLE_ERROR',
  FETCH_ARTICLE_SUCCESS: 'FETCH_ARTICLE_SUCCESS',
  HIGHLIGHT_UPLOAD_SUCCESS: 'HIGHLIGHT_UPLOAD_SUCCESS',
  HIGHLIGHT_UPLOADING: 'HIGHLIGHT_UPLOADING',
  HIGHLIGHT_UPLOAD_FAILED: 'HIGHLIGHT_UPLOAD_FAILED',
};
