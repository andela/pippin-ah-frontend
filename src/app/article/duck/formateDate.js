export default rawDate => {
  const date = new Date(rawDate).toDateString();
  return {
    weekDay: date.slice(0, 3),
    month: date.slice(4, 7),
    day: date.slice(8, 10),
    year: date.slice(11, 15),
  };
};
