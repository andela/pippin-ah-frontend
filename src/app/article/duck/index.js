export {
  doCreateArticle,
  doFetchArticle,
  doSetCategory,
  doUpdateCategoryData,
  doFetchArticles,
  doBookmarkArticle,
  doRemoveBookmark,
} from './operations';
export { createArticleReducer, fetchArticleReducer } from './reducers';
export { default as constants } from './constants';
export { default as types } from './types';
export { default as actions } from './actions';
export { default as formatDate } from './formateDate';
