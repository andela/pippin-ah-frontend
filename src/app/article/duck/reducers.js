import types from './types';

const INITIAL_STATE = {
  createStatus: {
    status: '',
    data: '',
  },
  singleFetchStatus: {
    status: '',
    data: '',
  },
  fetchArticleStatus: {
    fetchArticleState: '',
    errorMessage: '',
  },
  bookmarkArticleStatus: {
    bookmarkArticleState: '',
    errorMessage: '',
  },
  articleData: {
    Arts: [],
    Mathematics: [],
    Science: [],
    Engineering: [],
    Technology: [],
    Bookmarks: [],
  },
};

const createArticleReducer = (state = INITIAL_STATE, action) => {
  switch (action.type) {
    case types.SET_CREATE_STATUS: {
      const { createStatus } = action;
      return {
        ...state,
        createStatus,
      };
    }
    case types.SET_SINGLE_FETCH_STATUS: {
      const { singleFetchStatus } = action;
      return {
        ...state,
        singleFetchStatus,
      };
    }
    default:
      return state;
  }
};

const fetchArticleReducer = (state = INITIAL_STATE, action) => {
  switch (action.type) {
    case types.SET_FETCH_ARTICLE_STATE: {
      const { fetchArticleState } = action;
      return {
        ...state,
        fetchArticleState,
      };
    }

    case types.SET_FETCH_ARTICLE_ERROR: {
      const { errorMessage } = action;
      return {
        ...state,
        errorMessage,
      };
    }

    case types.SET_ARTICLE_CATEGORY: {
      const { articleCategory } = action;
      return {
        ...state,
        articleCategory,
      };
    }

    case types.SET_CURRENT_PAGE: {
      const { currentPage } = action;
      return {
        ...state,
        currentPage: { ...state.currentPage, ...currentPage },
      };
    }

    case types.ADD_ARTICLE_DATA: {
      const { articleData } = action;
      const oldArticleData = state.articleData;
      return {
        ...state,
        articleData: { ...oldArticleData, ...articleData },
      };
    }
    case types.ADD_NEWLY_CREATED_ARTICLE: {
      const { newArticle } = action;
      const currentArtcleData = state.articleData;
      const newArticleCategory = newArticle.category;
      const articleToAdd = {
        [newArticleCategory]: [
          newArticle,
          ...currentArtcleData[newArticleCategory],
        ],
      };
      return {
        ...state,
        articleData: { ...currentArtcleData, ...articleToAdd },
      };
    }

    case types.UPDATE_CATEGORY_DATA: {
      const { appendedCategoryData } = action;
      const oldArticleData = state.articleData;
      return {
        ...state,
        articleData: { ...oldArticleData, ...appendedCategoryData },
      };
    }

    default:
      return state;
  }
};

const bookmarkArticleReducer = (state = INITIAL_STATE, action) => {
  switch (action.type) {
    case types.SET_BOOKMARK_ARTICLE_STATE: {
      const { bookmarkArticleState } = action;
      return {
        ...state,
        bookmarkArticleState,
      };
    }
    case types.SET_BOOKMARK_ARTICLE_ERROR: {
      const { errorMessage } = action;
      return {
        ...state,
        errorMessage,
      };
    }
    default:
      return state;
  }
};

export { createArticleReducer, fetchArticleReducer, bookmarkArticleReducer };
