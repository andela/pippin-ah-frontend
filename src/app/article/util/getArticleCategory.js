const getArticleCategory = () => {
  const category = window.location.pathname.split('/')[2];
  const titleCasedCategory = category
    ? category.replace(/^[a-z]/, x => x.toUpperCase())
    : '';
  return titleCasedCategory;
};

export default getArticleCategory;
