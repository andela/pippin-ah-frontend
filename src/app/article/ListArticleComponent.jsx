import React, { Fragment } from 'react';
import lifecycle from 'react-pure-lifecycle';
import { Link } from 'react-router-dom';
import { constants } from './duck';
import getArticleCategory from './util/getArticleCategory';
import { EllipsisLoaderComponent } from '../loaders';
import './ListArticle.scss';

const category = getArticleCategory();

/* istanbul ignore next */
const methods = {
  componentDidMount({ fetchArticle, articleData, setCategory }) {
    if (!articleData) {
      fetchArticle(category);
    }
    if (articleData && articleData[category]) {
      return setCategory(category);
    }
  },
  componentDidUpdate({
    articleCategory,
    articleData,
    fetchArticle,
    setCategory,
  }) {
    const newCategory = getArticleCategory();
    const storeData = articleData && articleData[newCategory];
    if (articleCategory !== newCategory && !storeData) {
      fetchArticle(newCategory);
    }
    return setCategory(newCategory);
  },
};

const elipsisLoader = (
  <>
    <div className="li-loader">
      <EllipsisLoaderComponent />
    </div>
  </>
);

const ListArticleComponent = ({
  fetchArticleState,
  articleData,
  articleCategory,
}) => {
  return (
    <Fragment>
      <div id="liArticleContainer" className="container">
        {fetchArticleState === 'FETCHING_ARTICLE' && elipsisLoader}
        <>
          <div className="col s12 center-align">
            <h3>{articleCategory}</h3>
          </div>
          <div className="col s12 right-align write-article">
            <Link to="/create-article">Write Article</Link>
          </div>
        </>
        <div className="row">
          {/* istanbul ignore next */
          fetchArticleState === constants.FETCH_ARTICLE_SUCCESS &&
            articleData[articleCategory] &&
            articleData[articleCategory].map(article => (
              <div
                id="liMainDivXz"
                className="col s12 m6 l4"
                key={article.slug}
              >
                <div className="card" id="liCardDivXz">
                  <div id="articleImageDivXz">
                    <img
                      id="articleImageXz"
                      className="activator"
                      alt="Cover"
                      src={article.coverImageUrl}
                    />
                  </div>
                  <div id="listTitleDivXz" className="card-content right-align">
                    <span
                      id="listTitleXz"
                      className="card-title activator flow-text truncate center-align"
                    >
                      {article.title}
                    </span>
                    <strong>by: {article.author}</strong>
                    <div id="liAuthorNameXz">
                      <Link to={`/article/${article.slug}`}>Read more</Link>
                    </div>
                  </div>
                  <div className="card-reveal">
                    <span className="card-title grey-text text-darken-4">
                      {article.title}
                      <i className="material-icons right">close</i>
                    </span>
                    <p>{article.description}</p>
                    <div>
                      <p>
                        <span id="readTime">Read time: </span>
                        {article.readTime}
                        {article.readTime === 1 ? ' minute' : ' minutes'}
                      </p>
                      <Link to={`/article/${article.slug}`}>
                        Click here to read
                      </Link>
                    </div>
                  </div>
                </div>
              </div>
            ))}
        </div>
      </div>
    </Fragment>
  );
};

export default lifecycle(methods)(ListArticleComponent);
