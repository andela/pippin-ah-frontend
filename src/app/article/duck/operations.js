import axios from 'axios';
import actions from './actions';
import constants from './constants';
import { uploadImage } from '../../util/uploadToCloudinary';

const url = 'https://learnground-api-staging.herokuapp.com/api/v1/articles';
const { setCreateStatus } = actions;
const doCreateArticle = articleDetails => dispatch => {

  if (!articleDetails.uploadCoverUrl) {
    return dispatch(
      setCreateStatus({
        status: constants.CREATE_ERROR,
        data: 'You must select a cover image',
      }),
    );
  }
  dispatch(setCreateStatus({ status: constants.CREATING }));
  const { uploadCoverUrl } = articleDetails;
  uploadImage('article', uploadCoverUrl)
    .then(imageLink => {
      articleDetails.coverImageUrl = imageLink;
      const headers = {
        headers: { Authorization: localStorage.getItem('token') },
      };
      return axios
        .post(url, articleDetails, headers)
        .then(({ data }) => {
          return dispatch(
            setCreateStatus({
              status: constants.CREATE_SUCCESS,
              data: data.slug,
            }),
          );
        })
        .catch(({ response }) => {
          return dispatch(
            setCreateStatus({
              status: constants.CREATE_ERROR,
              data: response.data.error,
            }),
          );
        });
    })
    .catch(() => {
      return dispatch(
        setCreateStatus({
          status: constants.CREATE_ERROR,
          data: 'error uploading cover image',
        }),
      );
    });
};

export default doCreateArticle;
