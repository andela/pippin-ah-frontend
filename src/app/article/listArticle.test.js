import React from 'react';
import thunk from 'redux-thunk';
import axios from 'axios';
import configureStore from 'redux-mock-store';
import { mount } from 'enzyme';
import { Provider } from 'react-redux';
import ListArticleComponent from './ListArticleComponent';
import { actions, constants, fetchArticleReducer, types } from './duck';
import getArticleCategory from './util/getArticleCategory';
import { EllipsisLoaderComponent } from '../loaders';
import {
  ListArticleContainer,
  mapDispatchToProps,
  mapStateToProps,
} from './ListArticleContainer';

jest.mock('axios');

const {
  setFetchArticleState,
  setFetchArticleError,
  setArticleCategory,
  addArticleData,
} = actions;

describe('ListArticleComponentAction', () => {
  it('it should set fetchArticle state', () => {
    const action = setFetchArticleState(constants.FETCHING_ARTICLE);
    expect(action).toEqual({
      type: types.SET_FETCH_ARTICLE_STATE,
      fetchArticleState: constants.FETCHING_ARTICLE,
    });
  });

  it('it should set fetchArticle error message', () => {
    const action = setFetchArticleError(constants.FETCH_ARTICLE_ERROR);
    expect(action).toEqual({
      type: types.SET_FETCH_ARTICLE_ERROR,
      errorMessage: constants.FETCH_ARTICLE_ERROR,
    });
  });

  it('it should set article category', () => {
    const action = setArticleCategory('Arts');
    expect(action).toEqual({
      type: types.SET_ARTICLE_CATEGORY,
      articleCategory: 'Arts',
    });
  });

  it('it should add article data', () => {
    const action = addArticleData('Data');
    expect(action).toEqual({
      type: types.ADD_ARTICLE_DATA,
      articleData: 'Data',
    });
  });

  it('should setup default state values', () => {
    const state = fetchArticleReducer(undefined, {
      type: '@@INIT',
    });
    expect(state.fetchArticleStatus).toEqual({
      fetchArticleState: '',
      errorMessage: '',
    });
  });
});

describe('ListArticleContainer', () => {
  it('should show initial state', () => {
    const initialState = {
      fetchArticle: {
        fetchArticleState: '',
        errorMessage: '',
      },
    };
    expect(mapStateToProps(initialState).fetchArticleState).toEqual('');
    expect(mapStateToProps(initialState).errorMessage).toEqual('');
  });

  it('should dispatch action', () => {
    const dispatch = jest.fn();
    mapDispatchToProps(dispatch).fetchArticle();
    expect(typeof dispatch.mock.calls[0][0]).toEqual('function');
  });
});

describe('fetchArticleReducers', () => {
  it('should setup default state values', () => {
    const state = fetchArticleReducer(undefined, {
      type: '@@INIT',
    });
    expect(state.fetchArticleStatus).toEqual({
      fetchArticleState: '',
      errorMessage: '',
    });
  });

  it('it should change the fetchArticle state', () => {
    const action = {
      type: types.SET_FETCH_ARTICLE_STATE,
      fetchArticleState: constants.FETCHING_ARTICLE,
    };
    const state = fetchArticleReducer(undefined, action);
    expect(state.fetchArticleState).toEqual(constants.FETCHING_ARTICLE);
  });

  it('it should change the fetch article error message', () => {
    const action = {
      type: types.SET_FETCH_ARTICLE_ERROR,
      errorMessage: 'failed to fetch',
    };
    const state = fetchArticleReducer(undefined, action);
    expect(state.errorMessage).toEqual(action.errorMessage);
  });

  it('it should change the fetch article error message', () => {
    const action = {
      type: types.SET_ARTICLE_CATEGORY,
      articleCategory: 'Arts',
    };
    const state = fetchArticleReducer(undefined, action);
    expect(state.articleCategory).toEqual(action.articleCategory);
  });

  it('it should add article data', () => {
    const action = {
      type: types.ADD_ARTICLE_DATA,
      articleData: { Arts: [1, 2, 3] },
    };
    const state = fetchArticleReducer(undefined, action);
    expect(state.articleData).toEqual(action.articleData);
  });
});

describe('Connected ListArticleComponent Component Dispatches Success', () => {
  const initialState = {
    fetchArticle: {
      fetchArticleState: '',
      errorMessage: '',
    },
    articleData: 'Data',
    articleCategory: 'Arts',
  };
  const mockStore = configureStore([thunk]);
  const store = mockStore(initialState);
  let wrapper;
  beforeEach(() => {
    const response = { data: 'fetchArticle successful' };
    axios.get.mockResolvedValue(response);
    wrapper = mount(
      <Provider store={store}>
        <ListArticleContainer />
      </Provider>,
    );
  });

  it('it should render the connected component', () => {
    expect(wrapper.find(ListArticleComponent).length).toEqual(1);
  });

  it('it should dispatch fetchArticle action', () => {
    const storeActions = store.getActions();
    const storeState = store.getState();
    console.log('---++-->*<--++---', storeActions);
    expect(storeActions[0].type).toEqual('SET_FETCH_ARTICLE_STATE');
    expect(storeState.articleData).toEqual('Data');
  });
});

describe('Connected ListArticleComponent Dispatches fetchArticle Error', () => {
  const initialState = {
    fetchArticle: {
      fetchArticleState: '',
      errorMessage: '',
    },
  };
  const mockStore = configureStore([thunk]);
  const store = mockStore(initialState);
  beforeEach(() => {
    const response = {
      response: { data: { error: 'invalid credentials' } },
    };
    axios.post.mockImplementation(() => Promise.reject(response));
    mount(
      <Provider store={store}>
        <ListArticleContainer />
      </Provider>,
    );
  });

  it('it should dispatch error action', () => {
    const storeActions = store.getActions();
    expect(storeActions[1].fetchArticleState).toEqual('FETCH_ARTICLE_ERROR');
  });
});

describe('Loader Component', () => {
  const state = {
    fetchArticle: {
      fetchArticleState: 'FETCHING_ARTICLE',
    },
  };
  const mockStore = configureStore([thunk]);
  const store = mockStore(state);
  beforeEach(() => {
    const response = {
      response: { data: { error: 'failed to fetch' } },
    };
    axios.post.mockImplementation(() => Promise.reject(response));
  });

  it('it should render the EllipsisLoaderComponent if making request', () => {
    const component = mount(
      <Provider store={store}>
        <ListArticleContainer />
      </Provider>,
    );
    expect(component.contains(<EllipsisLoaderComponent />)).toEqual(true);
  });
});

/* describe('Get article category function', () => {
  const { pathname } = window.location;
  beforeAll(() => {
    Object.defineProperty(window.location, 'pathname', {
      writable: true,
    });
    window.location.pathname = jest.fn();
  });
  afterAll(() => {
    window.location.pathname = pathname;
  });
  it('should get article category', () => {
    window.location.pathname = '/articles/science';
    const result = getArticleCategory();
    console.log('---result', result);
    expect(result).toEqual('Science');
  })
  it('should return empty string', () => {
    window.location.pathname = '/';
    const result = getArticleCategory();
    console.log('result', result);
    expect(result).toEqual('');
  })
}) */
