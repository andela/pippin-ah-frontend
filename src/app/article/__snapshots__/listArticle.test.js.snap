// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Connected ListArticleComponent Component Dispatches Success it should render the connected component 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Connect(PureLifecycle(ListArticleComponent))
    articleData={
      Object {
        "Arts": Array [
          Object {
            "title": "first article",
          },
        ],
      }
    }
  >
    <PureLifecycle(ListArticleComponent)
      appendArticleData={[Function]}
      errorMessage=""
      fetchArticle={[Function]}
      fetchArticleState=""
      setCategory={[Function]}
    >
      <ListArticleComponent
        appendArticleData={[Function]}
        errorMessage=""
        fetchArticle={[Function]}
        fetchArticleState=""
        setCategory={[Function]}
      >
        <div
          className="container"
          id="liArticleContainer"
        >
          <div
            className="col s12 center-align"
          >
            <h3 />
          </div>
          <div
            className="row"
          />
        </div>
      </ListArticleComponent>
    </PureLifecycle(ListArticleComponent)>
  </Connect(PureLifecycle(ListArticleComponent))>
</Provider>
`;

exports[`fetchArticleReducers should render the component 1`] = `
<ListArticleComponent
  fetchArticle={
    [MockFunction] {
      "calls": Array [
        Array [
          "Science",
          1,
        ],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": undefined,
        },
      ],
    }
  }
/>
`;
